geoclustering/triclustering.py

```diff
@@ -0,0 +1,83 @@
+import concurrent.futures
+import logging
+
+from concurrent.futures import ThreadPoolExecutor
+
+from . import triclustering_numpy
+
+logger = logging.getLogger(__name__)
+
+
+class Triclustering(object):
+
+    def __init__(self, Z, nclusters_row, nclusters_col, nclusters_bnd,
+                 conv_threshold=1.e-5, max_iterations=1, nruns=1,
+                 epsilon=1.e-8):
+        """
+
+        :param Z:
+        :param nclusters_row:
+        :param nclusters_col:
+        :param nclusters_bnd:
+        :param conv_threshold:
+        :param niterations:
+        :param nruns:
+        :param epsilon:
+        """
+        self.Z = Z
+        self.nclusters_row = nclusters_row
+        self.nclusters_col = nclusters_col
+        self.nclusters_bnd = nclusters_bnd
+        self.conv_threshold = conv_threshold
+        self.max_iterations = max_iterations
+        self.nruns = nruns
+        self.epsilon = epsilon
+
+        self.client = None
+
+        self.row_clusters = None
+        self.col_clusters = None
+        self.bnd_clusters = None
+        self.error = None
+
+    def run_with_threads(self, nthreads=1):
+        """
+
+        :param nthreads:
+        :return:
+        """
+        with ThreadPoolExecutor(max_workers=nthreads) as executor:
+            futures = {
+                executor.submit(triclustering_numpy.triclustering,
+                                self.Z,
+                                self.nclusters_row,
+                                self.nclusters_col,
+                                self.nclusters_bnd,
+                                self.conv_threshold,
+                                self.max_iterations,
+                                self.epsilon):
+                r for r in range(self.nruns)
+            }
+            row_min, col_min, bnd_min, e_min = None, None, None, 0.
```

Comment:
`e_min` seem to be constantly 0?

Reply:
Great that you have caught this! Here I have implicitly assumed the error would be negative - which is not always the case, see your example. A better initialisation of `e_min` is required, not only for the triclustering, but for the coclustering as well. I'll open a new issue.

Reply:
I think the reason why we never encountered this issue is that I have always used integer values in the matrices, as taken from the test examples.